gpytorch/lazy/cat_lazy_tensor.py

```diff
@@ -185,7 +185,18 @@ def _get_indices(self, row_index, col_index, *batch_indices):
 
     def _getitem(self, row_index, col_index, *batch_indices):
         indices = [*batch_indices, row_index, col_index]
-        cat_dim_indices = indices[self.cat_dim]
+        cat_dim = self.cat_dim % len(self.shape)
+        for i, idx in enumerate(indices):
+            if idx is None:
+                cat_dim += 1
+                continue
+            if i == cat_dim:
+                break
+        cat_dim_indices = indices[cat_dim]
+        out_cat_dim = cat_dim - sum(isinstance(idx, int) for idx in indices[:cat_dim])
+        num_tensor_indices = sum(torch.is_tensor(idx) for idx in indices)
+        if num_tensor_indices > 0:
+            out_cat_dim -= num_tensor_indices - 1
```

Comment:
I *think* I understand what you're doing here, but it's not at all easy to follow. Could you please add some comments here to explain what's going on?

Reply:
The first part (189 - 194) is adjusting `cat_dim` to take into account unsqueezed dimensions. Line 196 is adjusting the `cat_dim` of the returned lazy cat tensor to take into account dimensions that were removed because of `int` indexing. And lines 197 - 199 adjust it for tensor indexes [1]: if `indices` contains `n > 0` tensor indices, those dimensions are removed and a dimension is added to the beginning [2].

... except turns out indexing is more complicated than I originally thought.
[1] Firstly, here I'm not handling the case when the shape of the tensor index has len() > 1. The number of dimensions added is actually the number of dimensions of (the broadcast of) all the tensor indices. 😱
[2] Apparently (this is really weird), if _all_ tensor indices are in adjacent dimensions (`int` indexing is a type of tensor indexing, just with shape `()`), the new dimensions are added to in place of the removed dimensions. If they're not adjacent, the new dimensions are added to the beginning: 😱😱
```python
import numpy as np
A = np.zeros((4,5,6,7,8))
idxs = [[0,0],[0,0],[0,0]]
print(A.shape)
# (4, 5, 6, 7, 8)
print(A[:,idxs,idxs,idxs,:].shape)
# (4, 3, 2, 8)
print(A[:,idxs,:,idxs,idxs].shape)
# (3, 2, 4, 6)
```

So, this will take more work to work properly. I'll leave adding comments until the end.